### ISSUE TITLE
BlazorWebView interaction with components

### ISSUE THREAD

**Issue by janseris** [dotnet/maui#5237]
### Description

Hi. I have tried out some BlazorWebView stuff now (Windows Forms) and also in May 2021 (WPF).
I though the problem would be over after 10 months but it persists.

What is the preferred way to communicate between WinForms/WPF and a component inside BlazorWebView?

Imagine I have a UserControl - a Windows Forms component which works using BlazorWebView.
I want to use this UserControl in multiple places at the same time. That means I have BlazorWebView for each place of use.
How do I get events from the component inside a specific BlazorWebView?

All the tutorials just show 1 BlazorWebView with 1 global AppState singleton service which stores int Counter from Counter component. This is not sufficient. 
What if I need to report e.g. SelectedItemChanged in a Blazor component inside the UserControl?

BlazorWebView contains RootComponent collection but the component instance cannot be retrieved from it afaik. There is just Selector and Type and Parameters but nothing like ComponentBase reference or anything.

In the image - how do I know _**which instance of the MyListComponent**_ which is a root component in BlazorWebView _**was clicked**_?

![BlazorWebView problem](https://user-images.githubusercontent.com/64279914/157914950-fb05caae-f3a6-4867-9a9c-57000aa8b74e.png)

Thank you

### Public API Changes

A way to interact with the root component of BlazorWebView, e.g. subscribe to its events, access its collections etc.

### Intended Use-Case

Please see attached image.

**Comment by Eilon** [comment]
Hi @janseris , in general the pattern is to use some shared state object and use Dependency Injection to ensure each Razor Component can access it, and other code in the rest of the app can also access it.

The small sample in this repo shows a little bit of that: https://github.com/dotnet/maui/blob/main/src/BlazorWebView/samples/BlazorWinFormsApp/Form1.cs

The `AppState` class is injected into some `*.razor` files, but there's also code in the WinForms main form class that access it to show the value of the counter.

**Comment by janseris** [comment]
> Hi @janseris , in general the pattern is to use some shared state object and use Dependency Injection to ensure each Razor Component can access it, and other code in the rest of the app can also access it.
> 
> The small sample in this repo shows a little bit of that: https://github.com/dotnet/maui/blob/main/src/BlazorWebView/samples/BlazorWinFormsApp/Form1.cs
> 
> The `AppState` class is injected into some `*.razor` files, but there's also code in the WinForms main form class that access it to show the value of the counter.

Thank you, I think you pointed me to the right direction!
If I understand it correctly, every UserControl should perform dependency injection on its own and have its own AppState defined by AddSingleton<AppState>(), am I right? 
I was doing it in a way that I perform DI only once when starting the Windows Forms application and then I am passing the IServiceCollection everywhere through the Windows Forms code.

What about business logic services? 
Can be services shared in the whole app for all Blazor components whose DI scope is inside a UserControl/Form DI scope?
E.g. service which is loading items but caches them. The application would have 1 singleton service for all UserControls because the cache shall be shared. But at the same time, each BlazorWebView with its wrapper should have its own AppState within its own DI scope. Is this possible?

What would be the meaning AddScoped scope here? I know that for WebAssembly, it is "user scope" vs Blazor Server it is "circuit scope".
I tried AddScoped<AppState> and it was null when invoking GetRequiredService<AppState>() method compared to when using AddSingleton<AppState>, it was not null. 
Is the behavior of "scoped services" (added via IServiceCollection.AddScoped<T>()) not defined when using BlazorWebView?

Thank you very much!

**Comment by Eilon** [comment]
@janseris yes with DI you generally set up the services in one "container" (that's the service collection), and then you consume those same services from multiple places in the app, such as from each BlazorWebView, or anywhere else in the app that needs those services. I think that in Blazor Hybrid apps there isn't much meaning to Scoped services because the entire app is always running, or it isn't running. It's mostly a choice between Singleton (one instance for the lifetime of the DI container), or transient (new instance each time it is resolved). You can still use the Scoped lifetime but I don't think it's very useful in a hybrid app.
